File: cpp/199. Binary Tree Right Side View.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// BFS, Time: O(n), Space: O(n)
public:
    vector<int> rightSideView(TreeNode* root) {
        if(!root) return vi();       
        queue<TreeNode*> q; q.push(root);
        vi ans;
        while(q.size()) {
            int qSize = q.size();
            for(int i = 0; i < qSize; i++) {
                TreeNode* node = q.front(); 
                q.pop();
                if(i==0) ans.push_back(node->val);
                if(node->right) q.push(node->right);
                if(node->left) q.push(node->left);
            }
        }
        return ans;
    }
};

#include<fmt/core.h>
#include<fmt/ranges.h>
using namespace fmt;
int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    TreeNode *l = new TreeNode(2);
    TreeNode *r = new TreeNode(3);
    TreeNode *root = new TreeNode(1, l, r);
    Solution solution;
    vi ans = solution.rightSideView(root);
    print("{}\n", ans);
}

File: python/199-Binary Tree Right Side View.py
from typing import List
from bintree_helper import *

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        "BFS, Time: O(n), Space: O(n)"
        if root is None: return []
        from collections import deque
        q = deque([root])
        result = []
        while q:
            sz = len(q)
            for i in range(sz):
                node = q.popleft()
                if i == 0:
                    result.append(node.val)
                if node.right: q.append(node.right)
                if node.left: q.append(node.left)
        return result
    def rightSideView(self, root: TreeNode) -> List[int]:
        "DFS, Time: O(n), Space: O(n)"
        result = []
        def dfs(root, level):
            if root is None: return
            if len(result) == level:
                result.append(root.val)
            if root.right: dfs(root.right, level+1)
            if root.left: dfs(root.left, level+1)
        dfs(root, 0)
        return result

root = deserialize("[1,2,3,null,5,null,4]")
result = Solution().rightSideView(root)
assert result == [1, 3, 4]